### DISCUSSION TITLE
issue: reset in useEffect is not working as expected in a Child Component

### DISCUSSION THREAD

**Discussion by LydiaF** [react-hook-form/react-hook-form#8532] [Q&A] [answered]
### Version Number

7.32.2

### Codesandbox/Expo snack

https://codesandbox.io/s/purple-grass-r2o3o5?file=/src/TestForm.js

### Steps to reproduce

Dear Bill,

I am trying to prepopulate a form while keeping the defaultValues plain so I can have isDirty = true. (Same goal as [this issue](https://github.com/react-hook-form/react-hook-form/discussions/7654).)

I apologise if I am missing something completely basic here, but I can't see why reset is not working if you run it in an effect in `<Child />`, whereas it does work if it is run in `<TestForm />` as expected.

```jsx
import React, { useEffect } from "react";
import { useForm } from "react-hook-form";

const Child = ({ reset, register }) => {
  // This does not work

  useEffect(() => {
    reset({ first_name: "hello" }, { keepDefaultValues: true });
  }, [reset]);

  return (
    <div>
      First Name:
      <input {...register("first_name")} />
    </div>
  );
};
export const TestForm = () => {
  const { register, reset } = useForm({ defaultValues: { first_name: "" } });

  // This will work

  // useEffect(() => {
  //   reset({ first_name: "hello" }, { keepDefaultValues: true });
  // }, [reset]);

  return (
    <div>
      <form>
        <Child reset={reset} register={register} />
      </form>
    </div>
  );
};

export default TestForm;
```

Thank you so much for any advice,

Lydia

### Expected behaviour

The string "hello" would appear in the input.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment, marked as answer]
Hey @LydiaF this is expected, as the execution order of useEffect will execute at the child component before hitting, the parent one, and the subscription starts to kick in at `useEffect` `useForm` level. 

```tsx
function Child() {
  useEffect(() => {
     // run first before even useForm useEffect is fired.
     // order 1
  }, [])
  return null;
}

function Test() {
  // order 2
  useForm()
  useEffect()
  
  return <Child />
}

````

**Reply by LydiaF** [reply]
Dear Bill,

Thank you so much for your reply and replying so quickly. I was not aware of this kind of thing unfortunately.

Is there any way to achieve what I am trying to do? I would have kind of thought that the effect would have run again with a new reset function when the subscription kicks in but obviously that is not happening.

Thank you!!

**Reply by bluebill1049** [reply]
Probably make sense to flush `reset` at the root of (useForm) level of your app, or you have to wait until useForm (mounted/useEffect kicked in)

**Reply by LydiaF** [reply]
I am really grateful for your help and your fantastic package. Thank you. We are good now. Was then getting caught by having shouldUnregister: true but was able to remove so no problem.
